sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnInvokers.java

```diff
@@ -0,0 +1,631 @@
+            .method(ElementMatchers.named("invokeProcessElement"))
+            .intercept(new ProcessElementDelegation(signature.getProcessElement()))
+            .method(ElementMatchers.named("invokeStartBundle"))
+            .intercept(new BundleMethodDelegation(signature.getStartBundle()))
+            .method(ElementMatchers.named("invokeFinishBundle"))
+            .intercept(new BundleMethodDelegation(signature.getFinishBundle()));
+
+    DynamicType.Unloaded<?> unloaded = builder.make();
+
+    @SuppressWarnings("unchecked")
+    Class<? extends DoFnInvoker<?, ?>> res =
+        (Class<? extends DoFnInvoker<?, ?>>)
+            unloaded
+                .load(DoFnInvokers.class.getClassLoader(), ClassLoadingStrategy.Default.INJECTION)
+                .getLoaded();
+    return res;
+  }
+
+  private abstract static class MethodDelegation implements Implementation {
+    FieldDescription delegateField;
+
+    @Override
+    public InstrumentedType prepare(InstrumentedType instrumentedType) {
+      // Remember the field description of the instrumented type.
+      delegateField =
+          instrumentedType
+              .getDeclaredFields()
+              .filter(ElementMatchers.named(FN_DELEGATE_FIELD_NAME))
+              .getOnly();
+
+      // Delegating the method call doesn't require any changes to the instrumented type.
+      return instrumentedType;
+    }
+
+    @Override
+    public ByteCodeAppender appender(final Target implementationTarget) {
+      return new ByteCodeAppender() {
+        @Override
+        public Size apply(
+            MethodVisitor methodVisitor,
+            Context implementationContext,
+            MethodDescription instrumentedMethod) {
+          StackManipulation manipulation =
+              isRequired()
+                  ? new StackManipulation.Compound(
+                      // Push "this" reference to the stack
+                      MethodVariableAccess.REFERENCE.loadOffset(0),
+                      // Access the delegate field of the the invoker
+                      FieldAccess.forField(delegateField).getter(),
+                      invokeTargetMethod(instrumentedMethod))
+                  : MethodReturn.VOID;
+          StackManipulation.Size size = manipulation.apply(methodVisitor, implementationContext);
+          return new Size(size.getMaximalSize(), instrumentedMethod.getStackSize());
+        }
+      };
+    }
+
+    protected abstract boolean isRequired();
+
+    protected abstract StackManipulation invokeTargetMethod(MethodDescription instrumentedMethod);
+  }
+
+  private static final class ProcessElementDelegation extends MethodDelegation {
+    private static final Map<DoFnSignature.ProcessElementMethod.Parameter, MethodDescription>
+        EXTRA_CONTEXT_FACTORY_METHODS;
+
+    static {
+      try {
+        Map<DoFnSignature.ProcessElementMethod.Parameter, MethodDescription> methods =
```

Comment:
It seems unnecessary to generalize here currently. I'd like to avoid structuring the code so that all parameters to all methods can only be exposed via ExtraContextFactory methods, because this might not be the case in the future (and it probably won't be, for some of the SplittableDoFn methods).